Micro-optimisations for matmul

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64387

A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Dispatch to `mv` rather than `mm` in the case `(n,) x (n, m)`.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of this, while fixing the CI and after further discussions, the following features were added. 
- [Fix CI] Add forward AD to `mv`, `matmul`, `__rmatmul__`
- [Fix CI] Add support for Half for a number of operations
- [Fix CI] Change some calls that called directly into the matmul implementation. Now they call into `_out`.
- Remove the uses of `set_`. (requested by @ezyang )
- Solve the resize bug in `matmul_out`.

Looking into the future, further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

Fixes https://github.com/pytorch/pytorch/issues/67767

cc @VitalyFedyunin @ngimel @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @heitorschueroff

lezcano committed:
Micro-optimisations for matmul

A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano

[ghstack-poisoned]

ngimel commented on test/test_linalg.py:
```diff
@@ -4956,98 +4956,111 @@ def test_triangular_solve_out_errors_and_warnings(self, device, dtype):
+                k = x.shape[-1]  # Scale the atol with the size of the matrix
+                if k == 0:
+                    k = 1
+                self.assertTrue(np.allclose(ans_numpy, expected, atol=k * 5e-6))
```
why doesn't self.assertEqual work here?

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

Lezcano commented:
The following script:
```python
from torch.utils.benchmark import Timer
import timeit
timer = Timer("torch::matmul(m1, m2);", setup="at::Tensor m1=torch::zeros({1,1,1});at::Tensor
m2=torch::zeros({1,1,1});", language="c++", timer=timeit.default_timer)
stats=timer.collect_callgrind(number=30, repeats=3)
print(stats[1].as_standardized().stats(inclusive=False))
```

Throws an instruction count of `951003` in master and of `872341` in this PR.

Using shapes `(1, 1, 1)` and `(1, 1)` it goes from `472891` (master) to `459918` (this PR).

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

Lezcano commented:
A more thorough instruction bench.
| Shapes | master  | This PR |
| --- | ------------- | ------------- |
| (1,1,1), (1,1,1) | 951003 | 870151 |
| (1,1,1), (1,1) | 472891  | 457008 |
| (1,1,1), (1,) | 585042  | 426104 |
| (1,1), (1,1,1) | 871951 | 791223 |
| (1,), (1,1,1) | 1020331 | 566204 |
| (1,1), (1,1) | 251204  | 249914 |
| (1,1), (1,) | 222644 | 221354 |
| (1,), (1,1) | 436218 | 354948 |
| (1,), (1,) | 195584 | 194324 |

Note in particular the 50% instructions in the case when the rhs is a batch of matrices and the lhs is a vector.

The reduction in the case when we use the out version is likely higher, but I did not bench that branch as it's not used nearly as much as the main branch.

This PR implements a number of general principles to try to save on unnecessary computations:
- Make the code const-correct
- Prefer `DimVector` over `std::vector`
- Perform less operations on `DimVector`
- Call `mv` rather than `mm` whenever possible
- Call `native::mv_out` to avoid a dispatch
- Work with the optional value from `out` directly to avoid a ref bump

This is ready for review @ngimel

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

ngimel commented on test/test_linalg.py:
```diff
@@ -4959,98 +4959,80 @@ def test_triangular_solve_out_errors_and_warnings(self, device, dtype):
+                                     make_arg(size_y, noncontiguous=nctg_y))
+
+    @dtypesIfCUDA(torch.float, torch.complex64)  # Integer matmul just supported on CPU
+    @dtypes(torch.int64, torch.float, torch.complex64)
```
Does this noticeably increase test time? Looks like you are running more tests than previously. I'm not sure that those `ranges` are needed, maybe step of 2 will do if test time is a problem? They also don't test what #20452 was complaining about, which is 0-sized inputs.

Lezcano replied:
Zero sizes are tested as the ranges start at zero. I'll add also tests for the empty size.

The tests take about 40s in my box, but the tests before just tested longs on cpu, so this is expected. I think we could bruteforce sizes 0-4 and batches 0-3? Would that be a fair compromise?

ngimel replied:
Oh right, I was looking at the previous test where ranges start at 1. Yeah, sizes 0-4, batches 0-3 should be enough.

ngimel reviewed (commented):
Sorry for delay with review, this looks great, I have small questions about testing.

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1442,115 +1442,118 @@ Tensor matmul(
+        const Tensor output = at::_unsafe_view(at::mm_out(*out_opt, t1, tensor2), output_size);
+        return out_opt->set_(output);
+      } else {
+        const Tensor output = at::_unsafe_view(at::native::mv_out(t1, tensor2, *out_opt), output_size);
```
I know... There is no native::mm_out so I could not call it directly...

Lezcano commented on test/test_linalg.py:
```diff
@@ -4959,98 +4959,80 @@ def test_triangular_solve_out_errors_and_warnings(self, device, dtype):
+        assertEqual(ans, expected)
 
-        out = torch.zeros(*shape, dtype=torch.int64).to(x.device)
+        out = torch.empty((0,), dtype=x.dtype, device=x.device)
```
Luckily we compute the non-out version just before, so we have the shapes there :)

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

Lezcano commented:
@ngimel There are some tests that are failing on the integer matrices with errors of the form:
>AssertionError: False is not true : Tensors failed to compare as equal!Found 1 different element(s) (out of 1), with the greatest difference of 94640972488000 (94640972488000 vs. 0) occuring at index (0, 0, 0, 0).

This is a bit odd, as the only tests that were there before this PR were those for the `int64` type. Also, these tests pass in my local box, so I cannot reproduce this failure. Could you give me a hand with this?

ngimel commented:
That's weird, can you add printouts to see at which shapes this is happening, and which path in the matmul is taken? Sorry we don't have ssh into failing jobs, so it's debugging by CI :-(

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1452,115 +1465,114 @@ Tensor matmul(
 
   if (dim_tensor1 == 1 && dim_tensor2 == 1) {
-    return has_out ? at::native::dot_out(tensor1, tensor2, out) : tensor1.dot(tensor2);
+    return has_out ? at::dot_out(tensor1, tensor2, *out_opt) : tensor1.dot(tensor2);
```
I'm confused about this (and the other `_out` below). Isn't this function a composite implicit autograd?

Lezcano replied:
Yes, but note that if out is defined, then it calls `dot_out`, otherwise it does the dispatch. @ngimel told me to remove the `native::`, but I really don't see why this should be removed in this path.

Lezcano replied:
(at least not from an autograd perspective. I have no idea about tracing and all that...)

albanD replied:
Ho that's only when matmul is called with `out` ok!
sounds great!

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

Lezcano commented:
I updated the PR as a few tests were failing.

The good part is that I added support for forward AD and `torch.float16` for a number of linear operations

The bad part is that an assert in the sampling function of `det` was not robust enough and started failing. cc @nikitaved who worked on this function. The test for the singularity of `det` are a bit too brittle and keep failing any time something changes. We should consider either removing them or somewhat relasing the `atol` / `rtol` in those gradchecks.

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

Lezcano commented:
@JackCaoG There's an XLA test failing, namely `AtenXlaTensorTest.TestMatmul_1x2`, but I have not been able to find it in the XLA repo. Could you point me to it?

Also, how brittle are these tests? It may be the case that it is failing because I moved that operation from using `mm` to using `mv`. See
https://github.com/pytorch/pytorch/blob/09a5db65a29459944359d36c3becdc5df06dbdae/aten/src/ATen/native/LinearAlgebra.cpp#L1490-L1492

Does XLA need to be updated to account for these changes?

JackCaoG commented:
@Lezcano I think it is in [here](https://github.com/pytorch/xla/blob/master/test/cpp/test_aten_xla_tensor.cpp#L3449), I will take a look .

JackCaoG commented:
@Lezcano I submit a pytorch/xla pr to fix the test issue. PyTorch/XLA needs https://github.com/pytorch/pytorch/pull/66018 present in this branch to build successfully,  could you rebase this pr before merging? This way I can make sure all test passed on pt/xla pr side.

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel heitorschueroff jianyuh nikitaved pearu mruberry walterddr xwang233 @Lezcano

[ghstack-poisoned]

Lezcano commented:
@pytorchbot ciflow rerun

Lezcano commented:
@ngimel This PR should be almost ready to land. All the errors but seem unrelated. There's just one that comes from a flaky test for `det` that keeps breaking fairly often. I was thinking of disabling that one because of all the noise it creates. Should I do so in this PR, or should I stack one on top of this one that just disables that test?

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.

This also includes an optimisation by dispatching vector matrix prodcuts
to `mv` rather than `mm`.

Further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano heitorschueroff

[ghstack-poisoned]

Lezcano commented:
Re. The test failures.
- The one from XLA is real, but  @JackCaoG's fix will fix it after landing.
- One is related to `det`'s flaky tests. We should disable these flaky tests.
- The one from [distributed](https://github.com/pytorch/pytorch/pull/64387/checks?check_run_id=3830708875) is really odd. I have also checked the 4 places where `cuda::blas::gemv` is called from in the codebase, and all of them make sure that `incx` is set to something larger than zero. I cannot ssh into the test, nor I have been able to run the test locally. I can't repro running parts of the tests on my box either, so I don't know how to tackle this one. 
- The one from [mobile](https://app.circleci.com/pipelines/github/pytorch/pytorch/391079/workflows/a87f25c1-6e1b-4d59-8eaf-dfbd9cb3e5c0/jobs/16370530) seems unrelated, but it might be related to having to register `mv` for mobile somewhere. I don't really know.

Help with any of the last two would be very much appreciated.

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Dispatch to `mv` rather than `mm` in the case `(n,) x (n, m)`.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of this, while fixing the CI and after further discussions, the following features were added. 
- [Fix CI] Add forward AD to `mv`, `matmul`, `__rmatmul__`
- [Fix CI] Add support for Half for a number of operations
- [Fix CI] Change some calls that called directly into the matmul implementation. Now they call into `_out`.
- Remove the uses of `set_`. (requested by ezyang )
- Solve the resize bug in `matmul_out`.

Looking into the future, further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc @VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Dispatch to `mv` rather than `mm` in the case `(n,) x (n, m)`.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of this, while fixing the CI and after further discussions, the following features were added. 
- [Fix CI] Add forward AD to `mv`, `matmul`, `__rmatmul__`
- [Fix CI] Add support for Half for a number of operations
- [Fix CI] Change some calls that called directly into the matmul implementation. Now they call into `_out`.
- Remove the uses of `set_`. (requested by ezyang )
- Solve the resize bug in `matmul_out`.

Looking into the future, further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Micro-optimisations for matmul"


A number of these include:

- Always prefer `DimVector` over `std::vector` when handling dimensions.
- Make the code `const` correct.
- Create `DimVector`'s more efficiently (e.g. prefer `append` over
`insert`).
- Access sizes of the tensors via `sizes().front()` / `sizes().back()`
  / `sizes().end()[-2]`
- Do not create intermediary tensors / vectors when it can be avoided.
- Dispatch to `mv` rather than `mm` in the case `(n,) x (n, m)`.
- Call `reshape` rather than `expect_contiguous`  + `view`

On top of this, while fixing the CI and after further discussions, the following features were added. 
- [Fix CI] Add forward AD to `mv`, `matmul`, `__rmatmul__`
- [Fix CI] Add support for Half for a number of operations
- [Fix CI] Change some calls that called directly into the matmul implementation. Now they call into `_out`.
- Remove the uses of `set_`. (requested by ezyang )
- Solve the resize bug in `matmul_out`.

Looking into the future, further optimisations could include dispatching to lower level functions
when one of the inputs has shape `(n, 1)` or `(1, n)`.

cc VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano heitorschueroff

[ghstack-poisoned]

Lezcano commented:
@iseeyuan this PR breaks a mobile build with the following error.
I reckon it could be caused because one or some of the utils that this PR introduces in `matmul` are not being compiled for mobile? Do you have any idea on how to approach this?
Thanks!
```+ run_predictor
+ cd /var/lib/jenkins/workspace/build_test_custom_build/predictor
+ ./Predictor /var/lib/jenkins/workspace/build_test_custom_build/MobileNetV2.pt
terminate called after throwing an instance of 'c10::Error'
  what():  false"TYPE" CHECK FAILED at "../../aten/src/ATen/core/dispatch/OperatorEntry.cpp"
Exception raised from reportError at ../../aten/src/ATen/core/dispatch/OperatorEntry.cpp:434 (most recent call first):
frame #0: ./Predictor() [0x1d00179]
frame #1: ./Predictor() [0x57d29a]
frame #2: ./Predictor() [0xb6892d]
frame #3: ./Predictor() [0xb511a9]
frame #4: ./Predictor() [0x1aaedf2]
frame #5: ./Predictor() [0x1aafea8]
frame #6: ./Predictor() [0x1aaff78]
frame #7: ./Predictor() [0x1a794c4]
frame #8: ./Predictor() [0x1a72b85]
frame #9: ./Predictor() [0x1a761d7]
frame #10: ./Predictor() [0x1a76699]
frame #11: ./Predictor() [0x1a777e1]
frame #12: ./Predictor() [0x1a778ef]
frame #13: main + 0xb9 (0x44fd29 in ./Predictor)
frame #14: __libc_start_main + 0xf0 (0x7f3df1bf6840 in /lib/x86_64-linux-gnu/libc.so.6)
frame #15: _start + 0x29 (0x49d1d9 in ./Predictor)
```

Lezcano commented:
@JacobSzwejbka  could you have a look at this error from mobile as well? I have not idea what's going on here nor how to debug it even. cf. https://github.com/pytorch/pytorch/pull/64387#issuecomment-964907362

Lezcano commented:
I have to have another look at this one and split it into several smaller PRs to find where's the problem. I'll do that soon™

Lezcano commented:
@pytorchbot ciflow rerun

Lezcano commented:
@pytorchbot ciflow rerun -l ciflow/all

Lezcano commented:
This PR has been split into a number of PRs to try to isolate the CI errors. See https://github.com/pytorch/pytorch/pull/75193
